api/client/src/main/java/org/projectnessie/client/http/impl/BaseHttpRequest.java

```diff
@@ -27,65 +27,170 @@
 import org.projectnessie.client.http.HttpRequest;
+import org.projectnessie.client.http.HttpResponse;
 import org.projectnessie.client.http.RequestContext;
+import org.projectnessie.client.http.ResponseContext;
+import org.projectnessie.client.http.Status;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public abstract class BaseHttpRequest extends HttpRequest {
 
+  private static final Logger LOGGER = LoggerFactory.getLogger(BaseHttpRequest.class);
+
   private static final TypeReference<Map<String, Object>> MAP_TYPE =
       new TypeReference<Map<String, Object>>() {};
 
   protected BaseHttpRequest(HttpRuntimeConfig config, URI baseUri) {
     super(config, baseUri);
   }
 
-  protected boolean prepareRequest(RequestContext context) {
+  @Override
+  public HttpResponse executeRequest(Method method, Object body) throws HttpClientException {
+    URI uri = uriBuilder.build();
+    RequestContext requestContext = new RequestContextImpl(headers, uri, method, body);
+    ResponseContext responseContext = null;
+    RuntimeException error = null;
+    try {
+      prepareRequest(requestContext);
+      responseContext = processResponse(uri, method, body, requestContext);
+      return config.responseFactory().make(responseContext, config.getMapper());
+    } catch (RuntimeException e) {
+      error = e;
+      throw e;
+    } finally {
+      processCallbacks(requestContext, responseContext, error);
+      cleanUp(responseContext, error);
+    }
+  }
+
+  protected void prepareRequest(RequestContext context) {
     headers.put(HEADER_ACCEPT, accept);
 
     Method method = context.getMethod();
-    boolean postOrPut = method == Method.PUT || method == Method.POST;
-    if (postOrPut) {
+    if (method == Method.PUT || method == Method.POST) {
       // Need to set the Content-Type even if body==null, otherwise the server responds with
       // RESTEASY003065: Cannot consume content type
       headers.put(HEADER_CONTENT_TYPE, contentsType);
     }
 
-    boolean doesOutput = postOrPut && context.getBody().isPresent();
     if (!config.isDisableCompression()) {
       headers.put(HEADER_ACCEPT_ENCODING, ACCEPT_ENCODING);
-      if (doesOutput) {
+      if (context.doesOutput()) {
         headers.put(HEADER_CONTENT_ENCODING, GZIP);
       }
     }
 
-    if (!bypassFilters) {
-      config.getRequestFilters().forEach(a -> a.filter(context));
-    }
+    processRequestFilters(context);
 
     HttpAuthentication auth = this.auth;
     if (auth != null) {
       auth.applyToHttpRequest(context);
       auth.start();
     }
+  }
 
-    return doesOutput;
+  protected ResponseContext processResponse(
+      URI uri, Method method, Object body, RequestContext requestContext) {
+    try {
+      ResponseContext responseContext = sendAndReceive(uri, method, body, requestContext);
+      processResponseFilters(responseContext);
+      mimicUrlConnectionBehavior(method, responseContext, uri);
```

Comment:
An exception here will miss closing the `responseContext` it seems.

Reply:
Good catch! These 2 statements need to be moved into `executeRequest`.